java/common/rules/impl/basic_java_library_impl.bzl

```diff
@@ -150,6 +152,19 @@ def basic_java_library(
     )
     target = {"JavaInfo": java_info}
 
+    if native_libraries:
+        dependencies_cc_info = cc_common.merge_cc_infos(cc_infos = native_libraries)
+
+        # Native dependencies have the same semantics as
+        # `cc_library#implementation_deps`. We only want to propagate
+        # `Cc{Debug,Linking}Context` to libraries depending on this.
+        target["CcInfo"] = CcInfo(
+            linking_context = dependencies_cc_info.linking_context,
+            debug_context = dependencies_cc_info.debug_context(),
+        )
+    else:
+        target["CcInfo"] = CcInfo()
```

Comment:
As this is a common case, you can reduce retained heap by sharing a single empty instance.

Reply:
I'm not sure how much of a difference this makes in practice: `java_library` will output `CcInfo`, so every target with `deps` will take the `if` case and produce a different, non-equal `CcInfo` with probably no native deps.

Reply:
Hmm, I see. I don't know whether `CcInfo` merging attempts to preserve empty instances, which would help here.

That said, I now realize that I don't quite understand why each `java_library` now needs to advertise `CcInfo`. What information is missing from `transitive_native_libraries`? Why can't this be a private provider distinct from `CcInfo` up to the `java_binary` level?